async_asgi_testclient/tests/asgi_spec/conftest.py

```diff
@@ -0,0 +1,175 @@
+    INIT = 1
+    READY = 2
+    SHUTDOWN = 3
+
+
+class BaseMockApp(object):
+    """A mock application object passed to TestClient for the tests"""
+
+    # Make it easy to override these for lifespan related test scenarios
+    lifespan_startup_message = {"type": "lifespan.startup.complete", "message": "OK"}
+    lifespan_shutdown_message = {"type": "lifespan.shutdown.complete", "message": "OK"}
+    use_lifespan = True
+
+    def __init__(self, **kwargs):
+        for k, v in kwargs:
+            setattr(self, k, v)
+        self.state = AppState.PREINIT
+
+    async def lifespan_startup(self, scope, receive, send, msg):
+        if self.state == AppState.READY:
+            # Technically, this isn't explicitly forbidden in the spec.
+            # But I think it should not happen.
+            raise RuntimeError("Received more than one lifespan.startup")
+        self.state = AppState.READY
+        return await send(self.lifespan_startup_message)
+
+    async def lifespan_shutdown(self, scope, receive, send, msg):
+        if self.state == AppState.SHUTDOWN:
+            # Technically, this isn't explicitly forbidden in the spec.
+            # But I think it should not happen.
+            raise RuntimeError("Received more than one lifespan.shutdown")
+        self.state = AppState.SHUTDOWN
+        return await send(self.lifespan_shutdown_message)
+
+    async def lifespan(self, scope, receive, send):
+        if not self.use_lifespan:
+            raise RuntimeError(f"Type '{scope['type']}' is not supported.")
+        while True:
+            try:
+                msg = await receive()
+            except RuntimeError as e:
+                if e.args == ("Event loop is closed",):
+                    return
+                else:
+                    raise
+
+            if msg["type"] == "lifespan.startup":
+                await self.lifespan_startup(scope, receive, send, msg)
+            elif msg["type"] == "lifespan.shutdown":
+                await self.lifespan_shutdown(scope, receive, send, msg)
+            else:
+                raise RuntimeError(f"Received unknown message type '{msg['type']}")
+            if self.state == AppState.SHUTDOWN:
+                return
+
+    async def http_request(self, scope, receive, send, msg):
+        # Default behaviour, just send a minimal response with OK to any request
+        await send({"type": "http.response.start", "headers": [], "status": 200})
+        await send({"type": "http.response.body", "body": b"OK"})
+
+    async def http_disconnect(self, scope, receive, send, msg):
+        raise RuntimeError(f"Received http.disconnect message {msg}")
+
+    async def http(self, scope, receive, send):
+        msg = []
+        # Receive http.requests until http.disconnect or more_body = False
+        while True:
+            msg.append(await receive())
+            if msg[-1]["type"] == "http.disconnect" or not msg[-1].get(
+                "more_body", False
+            ):
```

Comment:
Should we check that if have not received a `http.disconnect`, it is a `http.request`?

Like:

```suggestion
            if msg[-1]["type"] == "http.disconnect" or (msg[-1] == "http.request" and not msg[-1].get(
                "more_body", False
            )):
```

So we can catch unexpected events sent by asgi-testclient.

Reply:
Yes this makes sense. That whole function could probably be rewritten to be more readable too, now that I look at it after a break. ;-)

Reply:
great!